### ISSUE TITLE
Add a user-defined rule timeout

### ISSUE THREAD

**Issue by Gifford47** [openhab/openhab-core#3127]
In relation to this [article](https://community.openhab.org/t/why-have-my-rules-stopped-running-why-thread-sleep-is-a-bad-idea/47695) , a total of only 5 threads are allowed for the execution of the rules. The rules can be interrupted or stopped in many ways. For example, if Things or items are not updated as expected, the rest of a rule's code may stop executing. For the user it is then difficult to understand what and why exactly the system seems to "hang". Important rules (e.g. alarm systems) can also no longer be executed.

Therefore it would make sense to add an optional, user defined timeout to the rule, which then terminates the rule with a corresponding error message and releases the thread again.

**Comment by Gifford47** [comment]
> That article is more than four years old and no longer applies to OH 3. In OH 3 each rule gets its own thread. There are also events generated when a rule starts running so you can see a badge in MainUI showing when a rule is running. If a single rule stops running because it's stuck (which I've never seen happen on OH 3) it will be apparent because that rule will be the only one not responding and it will show as constantly running in MainUI.
> 
> You can reset it be disabling the rule (pause button in MainUI) and reenabling it and you will just have the one rule to debug.

Ok, but that implies that the user have to check if the rule is still running. My suggestion was, that OH stops those dead threads  automatically to release the thread for other purposes.
The story behind this issue is, that my OH3.3 system started a new thread for the same rule, but the previously started rule blocked something. So the new started thread did nothing. The ui showed me that the rule was restarted, but no actions were performed. After starting, it then switched back to "idle". Only when I paused and restarted the rule did it work again.

**Comment by rkoshak** [comment]
> My suggestion was, that OH stops those dead threads automatically to release the thread for other purposes.

But the thread is never "dead". It's running. It might be taking a long time because of the way it's coded but it's not stuck forever unless you've somehow done something very weird to make that happen and you'd almost have to do that on purpose.

I've never seen a case where a rule just stops while running forever unless someone has coded it that way on purpose.

> The story behind this issue is, that my OH3.3 system started a new thread for the same rule, but the previously started rule blocked something

You need to provide a lot more detail. What rules language? Created a new thread how? Generally creating new threads is not permitted from rules and the proper way to run something "in the background" is to use a Timer (which runs in the same thread as the rule BTW). 

> The ui showed me that the rule was restarted, but no actions were performed.

I've never seen a "restarted" state for a rule. I've only ever seen "RUNNING", "IDLE" and "UNINITIALIZED (DISABLED)". A RESTARTED state implies something is wrong with the rule/rule engine that is independent of any code that you wrote to execute in that rule. Or you've mistyped/misremembered the state the rule was in. But if that's the case, you don't need a timeout for the rule thread. Just look for that error RESTARTED state.

But yes, if the rule is in some error state, or the rule is RUNNING, a new trigger of the rule won't do anything until the currently running rule completes.

This is not a trivial thing to implement. It will cause huge disruption to many parts of OH core. It's important to understand the actual root cause of your original problem rather than blindly making a huge change (potentially breaking) to OH that maybe doesn't even address your actual problem.

**Comment by Gifford47** [comment]
ok, thanks for the clarification! Here ist my rule that didn´t finish several times:

```
// Triggers:
// - When Harmony_activity received a command

// context: harmony-1
var harmony_stat = getThingStatusInfo("harmonyhub:hub:HarmonyHub").getStatus().toString()

infoscr_led_stripe.sendCommand("confirm")

// if a previous timer is running, cancel it
if( timer !== null ) {
    timer.cancel()
}

// wenn steckdose aus ist und Aktivität erfolgt und kein timer aktiv:
if( receivedCommand != "PowerOff" && tasmota_S20_4AD8EF_6383_Relay1_1.state == OFF && timer === null ) {
    logInfo("harmony.rules", "Starte Harmony (harmony offline),Command:" + receivedCommand.toString())
    tasmota_S20_4AD8EF_6383_Relay1_1.sendCommand(ON)
    timer = createTimer(now.plusSeconds(64), [|
        if( receivedCommand.toString().toLowerCase.contains("beamer") ) {
            beamer_lift.sendCommand(ON)            // fahre beamerlift runter
            Thread::sleep(7000)
        }
        // in case of missing the activity from harmony binding:
        for (i=0; i<no_retries_activity; i++) {
            logInfo("harmony.rules", "Starte Harmony activity:" + receivedCommand.toString())
            HarmonyHubHarmonyHub_CurrentActivity.sendCommand(receivedCommand.toString())
            Thread::sleep(4000)
            if( HarmonyHubHarmonyHub_CurrentActivity.state.toString() == receivedCommand.toString() ) {
                timer = null
                return
            }
        }
    ])
    //logInfo("harmony.rules", "DEBUG: cmd an harmony")
// wenn steckdose an ist und Aktivität erfolgt und kein timer aktiv:
} else if( receivedCommand != "PowerOff" && tasmota_S20_4AD8EF_6383_Relay1_1.state == ON && timer === null ) {
    logInfo("harmony.rules", "Starte Harmony (harmony online),Command:" + receivedCommand.toString())
    if(harmony_stat == "ONLINE"){
        if( receivedCommand.toString().toLowerCase.contains("beamer") ) {
            beamer_lift.sendCommand(ON)
            Thread::sleep(7000)
        }
        HarmonyHubHarmonyHub_CurrentActivity.sendCommand(receivedCommand.toString())
        Thread::sleep(7000)
        if( beamer_lift.state == ON && (false == receivedCommand.toString().toLowerCase.contains("beamer")) ) {
            beamer_lift.sendCommand(OFF)
        }
    }
// wenn alles ausgeschaltet werden soll:
} else if ( receivedCommand == "PowerOff" ){ //wenn harmony vorher an gewesen ist (also einen Status hatte), dann erst ausschalten! Sonst Schleife!
    logInfo("harmony.rules", "Schalte Harmony aus ...")
    //wenn harmony nicht bereits alles ausgeschaltet hat:
    if( HarmonyHubHarmonyHub_CurrentActivity.state.toString() != "PowerOff" ) {
        HarmonyHubHarmonyHub_CurrentActivity.sendCommand(receivedCommand.toString())
    }
    timer = createTimer(now.plusSeconds(25), [|
        if( beamer_lift.state == ON ) {
            beamer_lift.sendCommand(OFF)
            Thread::sleep(2000)
        }
        tasmota_S20_4AD8EF_6383_Relay1_1.sendCommand(OFF)
        timer = null
    ])
}
```

**Comment by rkoshak** [comment]
Except for the fact that the timers here take a very long time to run, I see nothing exceptional about this rule. A timeout wouldn't do anything in this case since the rule itself would only take a few hundred milliseconds to run. It's the timers which take a long time, and I was incorrect above. What I meant was that the timers have their own thread and I stated just the opposite.

Since the rule exits immediately the proposal here would do nothing. Your rule must have entered some sort of error state. But without further information about what state the rule really was in and logs to tell us what caused it I'm not sure what could be done. 

You can make your timers take less time by replacing the Thread::sleeps with timers.

```
    timer = createTimer(now.plusSeconds(64), [|
        var timeout = 0
        if( receivedCommand.toString().toLowerCase.contains("beamer") ) {
            beamer_lift.sendCommand(ON)            // fahre beamerlift runter
            timeout = 7
        }
        var retries = 0
        timer = createTimer(now.plusSeconds(timeout), [|
            if( retries > no_retries_activity 
                 || HarmonyHubHarmonyHub_CurrentActivity.state.toString() == receivedCommand.toString()) {
                timer = null
            }
            else {
                logInfo("harmony.rules", "Starte Harmony activity:" + receivedCommand.toString())
                HarmonyHubHarmonyHub_CurrentActivity.sendCommand(receivedCommand.toString())
                retries = retries + 1
                timer.reschedule(now.plusSeconds(4)
            }
        ])
    ])
```

**Comment by Gifford47** [comment]
OK, I will try to reproduce the same error and then post everything necessary. If this is not possible, I will change the timer as you described.

I suspect that the timer thread was never terminated because the condition ``HarmonyHubHarmonyHub_CurrentActivity.state.toString() == receivedCommand.toString()`` was never met.

this is also where my suggestion to insert a general timeout came from.

**Comment by Gifford47** [comment]
> I'm not sure I understand the question. You give it a time when you want it to execute. If you are asking for a timeout parameter on `createTimer` then I know of no such option.

that's exactly what i meant. from my point of view it would be helpful if the complete rule (with all its timer threads) would end after a certain timeout. i've seen various rules on the forum that wait for certain states of the items. if these states don't occur, the rule runs endlessly (as in my example).
in my opinion, i can't rewrite the rule, because i need this state, also for subsequent scenes.

I came to the function through homeassistant ([wait timeout](https://www.home-assistant.io/docs/scripts/#wait-timeout)), where a certain time is waited for a status, otherwise the rule is continued or cancelled.

In my opinion, such a function makes sense, especially when there is abnormal behaviour, e.g. when things do not update or a binding possibly has an error.

By the way: Are the timer threads also limited to a total of 5 or can I theoretically run 5 "rule threads" AND i.e. 12 timer threads (from different rules) simultaneously?

**Comment by rkoshak** [comment]
> that's exactly what i meant. from my point of view it would be helpful if the complete rule (with all its timer threads) would end after a certain timeout.

As discussed above, I don't think a timeout for rules is needed. For timers I could potentially see a use for that, but it would be independent of the rule.

There wouldn't be one timeout for both. And each timer would have its own. And just having done a deep dive into the code it does look like there is a default timeout already applied to Timers at 65 seconds.

>  i've seen various rules on the forum that wait for certain states of the items. if these states don't occur, the rule runs endlessly (as in my example).
> in my opinion, i can't rewrite the rule, because i need this state, also for subsequent scenes.

You are almost certainly incorrect on this. You almost certainly *can* rewrite the rule into event based rules, perhaps with some book keeping. It has always been the case that this sort of code that waits on a state change in an Item is an antipattern and definitely be be avoided. Even were one to add a timeout to the rule and timers, it *still* would be an antipattern to be avoided.

I took your most challenging part of your original rule above and showed how to implement it using Timers that start and exit quickly without sleeps.

> I came to the function through homeassistant ([wait timeout](https://www.home-assistant.io/docs/scripts/#wait-timeout)), where a certain time is waited for a status, otherwise the rule is continued or cancelled.

It will likely take some adjustment to move any rule you've written based on that pattern to something more event driven. In short, don't sit there blocking and waiting for an Item to change state. Set a flag and have another rule that triggers when the Items changes state. If the flag is set, do what you need to in that rule. Maybe set a timer or expire to indicate an error if that second rule doesn't happen in the right amount of time.

> By the way: Are the timer threads also limited to a total of 5 or can I theoretically run 5 "rule threads" AND i.e. 12 timer threads (from different rules) simultaneously?

Again, the 5 rules threads is no longer a thing. Each and every rule has its own dedicated thread. You will *never* run out of rules threads. The most you can do is starve out a single rule but that will never impact other rules. 

Timers, on-the-other-hand look like they might have a limitation of 5 threads in the pool. Though that should be configurable via a config file if I'm reading the code correctly but I can't see where it might be set. The threadpool name appears to be "scheduler", so let's assume the pool size is 5.

Then yes, if you have five long running Timers, it can be up to 65 seconds before one times out and exits and another Timer is allowed to run. 

It will never be the case that all your rules stop running. At most one rule will get blocked and, as far as I can tell, only blocked for up to 65 seconds at which point the next queued up trigger will run for up to 65 seconds, and so on.

It will never be the case that your Threads stop running forever. They might be blocked for up to 65 seconds if you have five Timers that are stuck. Even so, you should still write your rules to run to completion quickly and not sit there in a sleep for seconds and seconds.

So to answer your question, if you have N rules, you can have N rules running simultaneously. There is a one-to-one relationship between threads and rules. Each rule has exclusive access to its very own Thread.

It appears that there is the one global Timer threadpool. So that means up to five Timers across all your rules can be actively running at the same time. If that's not enough, it appears to be configurable (though I don't know how quite yet but the code is clear it can be configured). 

However, I could be misunderstanding things and it actually creates a separate thread pool for *each* rule (unlikely but I am not familiar with this code) in which case you can have up to five simultaneously running Threads *per rule*.  Based on how hard I had to push to get and keep the one-to-one thread to rule relationship I doubt this is the case.

tl;dr: 

- I'm even more convinced now that your original problem did not have anything to do with threads and thread pools.
- There is already a default timeout of 65 seconds on Timers and Rules (probably configurable).
- It is an antipattern to have rules or timers that take a long time (seconds) to run. These should be built around events and triggers, not sleeps.

I believe but can't confirm yet, if you time out an error is generated and put into the logs.

**Comment by Gifford47** [comment]
ok, thanks for the detailed answer. I have rewritten parts of the rule, but I personally think that a timeout for a rule would still make sense (regardless of the number of threads available, e.g. in order not to wait permanently for certain status queries as soon as one does not/ cannot work via an event trigger (binding issues, timing problems, workarounds etc.)).

Thanks for the deep-dive on the timer topic! the timeout has already been implemented here, so I'll close my issue, as all topics have already been explained in detail by rkoshak.
